### ISSUE TITLE
Cannot build compiler with the "full" version of extlib

### ISSUE THREAD

**Issue by andyli** [HaxeFoundation/haxe#9420]
For some unknown reason, I failed to build Haxe 4.1.0 in Debian unstable with the Debian packages.

The error I got:
```
File "libs/ttflib/tTFTools.ml", line 213, characters 27-31:
213 | 		let code = UCharExt.code code in
                                 ^^^^
Error: This expression has type UChar.uchar = UChar.t
       but an expression was expected of type UCharExt.t
```

According to the discussion in #8918, I believe it has something to do with case sensitivity. While I'm still searching for the root cause, it would be great if we can resolve it by renaming some files (probably `haxe-debian/libs/extlib-leftovers/uTF8.{ml,mli}`).

**Comment by Simn** [comment]
I don't remember why we even still have the UTF8 and UCharExt modules there because they should exist in extlib anyway...

Could you check if removing extlib from ttflib/dune fixes this for you?

**Comment by Simn** [comment]
Interestingly, my local extlib doesn't have these modules. The last file there is `std.mli`, but I have no idea why that is the case because the extlib repository clearly has uTF8.ml and such: https://github.com/ygrek/ocaml-extlib/blob/master/src/uTF8.mli

**Comment by RealyUniqueName** [comment]
I doubt Debian uses case-insensitive file system by default. So unless you've manually changed your file system, case sensitivity should not be the reason.

**Comment by Simn** [comment]
This is probably not about casing at all and a matter of `extlib` having `uTF8.ml`. From what I gather, that has moved to `extlib-compat`, so the problem could be that the build system has an older version of it.

**Comment by andyli** [comment]
> I doubt Debian uses case-insensitive file system by default. So unless you've manually changed your file system, case sensitivity should not be the reason.

It's a bit complicated on my end - I'm using Debian in WSL 2, and my work directory is in a NTFS drive. I do have case-sensitivity enabled for my directory, so I "think" it's case-sensitive and my simple manual testing agreed.

Even more strange, the error only occurs when I use the ocaml packages from Debian, but not when I use opam, even if I used all the libs with the same version numbers.

Maybe it has something to do with the compilation order and paths or whatever.

**Comment by andyli** [comment]
> Could you check if removing extlib from ttflib/dune fixes this for you?

I've just tried but it gave me the same error.

And somehow even if I removed both extlib and extlib-leftover and it also gave me the same error. I suppose it should error about some types not found?

**Comment by andyli** [comment]
FYI, here are the commands I used to reproduce the issue with docker.

```sh
docker run -it --rm debian:unstable bash
apt update
apt upgrade -y
apt install -y make git ca-certificates ocaml-nox ocaml-dune camlp5 libxml-light-ocaml-dev libsedlex-ocaml-dev libextlib-ocaml-dev libptmap-ocaml-dev libsha-ocaml-dev libmbedtls-dev zlib1g-dev libpcre3-dev neko
git clone --recursive https://github.com/HaxeFoundation/haxe.git
cd haxe
make
```

**Comment by andyli** [comment]
> Interestingly, my local extlib doesn't have these modules.

Looks like that's the reason why it succeeded when I use opam instead of the Debian extlib package.

The opam extlib used this, which excluded utf8 and uchar...
```
install: [ [make "minimal=1" "install"] ]
```

Not very sure what should be the right fix now....

**Comment by Simn** [comment]
What I don't understand is why you still get that error after removing `extlib` from the dune file. That suggests that it picks up a different `uTF8.mli` file from some other place.

**Comment by Simn** [comment]
Or not specifically .mli but whichever one of the 30 OCaml file extensions it was...

Anyway, removing `extlib` there should be good regardless because the lib doesn't need it.

**Comment by andyli** [comment]
> What I don't understand is why you still get that error after removing extlib from the dune file.

Maybe it's because ttflib depends on swflib, which depends on extlib?

I found a way to reproduce the issue using the opam packages: replace extlib with extlib-compat by just `opam install extlib-compat`.

I suggested the Debian extlib maintainer to build with `minimal=1` like the OPAM extlib package. But even it it solved the Debian case, I'm not sure about the other distros' extlib.

To be safe in the long run, would dune's default wrapping help here? i.e. removing `(wrapped false)` in extlib-leftover.

**Comment by RealyUniqueName** [comment]
Removing `(wrapped false)` from extlib-leftover breaks compilation:
```
<...>
File "libs/swflib/as3.mli", line 310, characters 24-36:
Error: Unbound module MultiArray
```

**Comment by andyli** [comment]
Probably need to change the references to `Extlib_leftovers.MultiArray` or something. I've no idea how the dune namespacing works though...

Another idea is to drop our uTF8 and/or uCharExt, and use [camomile](https://opam.ocaml.org/packages/camomile/).

**Comment by RealyUniqueName** [comment]
Fixed by #9645
